src/core/mantle/evaluator/simple/utils.ts

```diff
@@ -0,0 +1,92 @@
+import { JSONPath } from "jsonpath-plus";
+import { pick } from "lodash-es";
+
+import { StyleExpression } from "../../types";
+
+export function getCacheableProperties(styleExpression: StyleExpression, feature?: any) {
+  const properties = pick(feature, getCombinedReferences(styleExpression, feature));
+  return properties;
+}
+
+export function getCombinedReferences(expression: StyleExpression, feature?: any): string[] {
+  if (typeof expression === "string") {
+    return getReferences(expression, feature);
+  } else {
+    const references: string[] = [];
+    for (const [condition, value] of expression.conditions) {
+      references.push(...getReferences(condition, feature), ...getReferences(value, feature));
+    }
+    return references;
+  }
+}
+
+export function getReferences(expression: string, feature?: any): string[] {
```

Comment:
We have two function to parse expression. First is this function. Second is `replaceVariables`.
But these are very heavy task, so it will make app slower...
I think it's better to share the result of parsing with Expression class, and if we remove replaceVariables then performance will be more improved.

Reply:
I looks it'll be hard to separate out alot the repetitive logic and avoid iterating over the string again in `replaceVariables` now that i've removed JP related evaluations.